File: spring-core-lab-solutions/Lab-06-solution/petclinic/src/main/java/com/petclinic/tech/LogAspect.java
package com.petclinic.tech;

import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.util.StopWatch;

import java.util.concurrent.TimeUnit;

@Aspect @Component
public class LogAspect {
    private static final Logger logger = LoggerFactory.getLogger(LogAspect.class);

    @Before("execution(* com.petclinic..*Service.*(..))")
    public void logMethodsCalledBefore( JoinPoint joinPoint ) throws Throwable {
        String methodName = joinPoint.getSignature().getName();
        String className = joinPoint.getTarget().getClass().getName();
        logger.info("Before executing method: {}.{}", className, methodName);
    }

    @Around("execution(* com.petclinic..*Service.*(..))")
    public Object logMethodsCalled( ProceedingJoinPoint joinPoint ) throws Throwable {
        String methodName = joinPoint.getSignature().getName();
        String className = joinPoint.getTarget().getClass().getName();

        StopWatch stopWatch =  new StopWatch();
        stopWatch.start();

        Object result = joinPoint.proceed();
        stopWatch.stop();
        logger.info("time spent: {} ms", stopWatch.getTotalTime(TimeUnit.MILLISECONDS));
        return result;
    }

}

File: spring-core-lab-solutions/Lab-kotlin-solution/src/main/kotlin/com/petclinic/aspect/LogAspect.kt
package com.petclinic.aspect

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.springframework.stereotype.Component
import org.springframework.util.StopWatch

@Aspect @Component //@Profile("monitoring")
class LogAspect {
    @Around("execution(* com..*Service.*(..))")
    fun logTimeSpent(joinPoint: ProceedingJoinPoint): Any? {
        val stopWatch: StopWatch = StopWatch()
        stopWatch.start()
        println("inside LogAspect")
        val result =  joinPoint.proceed()
        stopWatch.stop()
        println("function called: ${joinPoint.signature}: time spent: ${stopWatch.totalTimeMillis} ms")
        return result

        // Log with Trace id: "1234"
        // call to Service A via message queue. Header: "12334"
        // rest call to Service B. Http header contains trace id "1234"
        // receives response from Service B
        // logs with Trace id: "1234"

    }
}